llvm/lib/Target/AIE/AIELegalizerHelper.cpp

```diff
@@ -153,17 +162,20 @@ bool AIELegalizerHelper::legalizeG_BUILD_VECTOR(LegalizerHelper &Helper,
 
   const AIEBaseInstrInfo *II = ST.getInstrInfo();
   MachineOperand *OperandBegin = MI.operands_begin(),
-                 *Operand = MI.operands_end() - 1;
-  while (Operand != OperandBegin) {
+                 *Operand = MI.operands_begin() + 1,
+                 *OperandEnd = MI.operands_end();
+  while (Operand != OperandEnd) {
     Register Reg = Operand->getReg();
     Register Dst = MRI.createGenericVirtualRegister(VecTy);
 
-    if (DstVecSize == 512 && Operand == OperandBegin + 1)
+    if (DstVecSize == 512 && Operand == OperandBegin + 1) {
       Dst = DstReg;
+    }
 
     // vpush takes 32-bit operands so we sign extend the input variable. This is
     // required here since we don't have 16 or 32-bit registers.
-    if (DstVecEltTy.getSizeInBits() != 32) {
+    if (DstVecEltTy.getSizeInBits() != 32 &&
+        MI.getOpcode() == TargetOpcode::G_BUILD_VECTOR) {
```

Comment:
Does that mean `G_BUILD_VECTOR_TRUNC` can accept `s8` and  `s16` types? I think we should always extend to 32-bits, because this is the size of our GPRs.

Reply:
Maybe I will add a check for input type `s32`